test/serviceContext-test.js

```diff
@@ -0,0 +1,329 @@
+    'request_object_signing_alg_values_supported': ['none', 'RS256', 'ES256'],
+    'display_values_supported': ['page', 'popup'],
+    'claim_types_supported': ['normal', 'distributed'],
+    'claims_supported': [
+      'sub', 'iss', 'auth_time', 'acr', 'name', 'given_name', 'family_name',
+      'nickname', 'profile', 'picture', 'website', 'email', 'email_verified',
+      'locale', 'zoneinfo', 'http://example.info/claims/groups'
+    ],
+    'claims_parameter_supported': true,
+    'service_documentation':
+        'http://server.example.com/connect/service_documentation.html',
+    'ui_locales_supported': ['en-US', 'en-GB', 'en-CA', 'fr-FR', 'fr-CA']
+  };
+
+  let res6 = verifyAlgSupport(ci, 'RS256', 'id_token', 'signing_alg');
+  it('verify_alg_support', function() {
+    assert.isTrue(res6);
+  });
+
+  let res7 = verifyAlgSupport(ci, 'RS512', 'id_token', 'signing_alg');
+  it('verify_alg_support', function() {
+    assert.isFalse(res7);
+  });
+
+  let res8 = verifyAlgSupport(ci, 'RSA1_5', 'userinfo', 'encryption_alg');
+  it('verify_alg_support', function() {
+    assert.isTrue(res8);
+  });
+
+  let res9 = verifyAlgSupport(ci, 'ES256', 'token_endpoint_auth', 'signing_alg');
+  it('verify_alg_support', function() {
+    assert.isTrue(res9);
+  });
+
+  ci.providerInfo['issuer'] = 'https://example.com/';
+  let url_list = ci.generateRequestUris('/leading');
+  let sp = urlParse(url_list[0]);
+  let p = sp.pathname.split('/');
+
+  it('verify_requests_uri', function() {
+    assert.deepEqual(p[0], '');
+    assert.deepEqual(p[1], 'leading');
+    assert.deepEqual(p.length, 3);
+  });
+
+  ci.providerInfo['issuer'] = 'https://op.example.org/';
+  url_list = ci.generateRequestUris('/leading');
+  sp = urlParse(url_list[0]);
+  let np = sp.pathname.split('/');
+
+  it('verify_requests_uri test2', function() {
+    assert.deepEqual(np[0], '');
+    assert.deepEqual(np[1], 'leading');
+    assert.deepEqual(np.length, 3);
+    assert.notDeepEqual(np[2], p[2]);
+  });
+});
+
+describe('client info tests', function() {
+  let config;
+  let ci;
+  beforeEach(function() {
+    config = {
+      'client_id': 'client_id',
+      'issuer': 'issuer',
+      'client_secret': 'client_secret',
+      'base_url': 'https://example.com',
+      'requests_dir': 'requests'
+    };
+
+    ci = new ServiceContext(null, config);
+  });
+
+  it('client info init', function() {
+    for (let i = 0; i < Object.keys(config); i++) {
+      let attr = Object.keys(config)[i];
+      if (attr === 'client_id') {
+        assert.deepEqual(ci.clientId, config[attr]);
+      } else if (attr === 'issuer') {
+        assert.deepEqual(ci.issuer, config[attr]);
+      } else if (attr === 'client_secret') {
+        assert.deepEqual(ci.clientSecret, config[attr]);
+      } else if (attr === 'base_url') {
+        assert.deepEqual(ci.base_url, config[attr]);
+      } else if (attr === 'requests_dir') {
+        assert.deepEqual(ci.base_url, config[attr]);
```

Comment:
Please do not use this kind of "for loop combined with if else if construct" ever again in unit tests. First of all it's really hard to read and furthermore it's error prone.

Like in this the tests pass even though they shouldn't as you `deepEqual` `ci.base_url` against `config['requests_dir']` which should return false based on the `config` used and defined above.  The root cause here is that this code path is never executed, because the for statement in line 269 is wrong. The for statement uses the keys array (`Object.keys(config)`) rather than the length of the keys array (`Object.keys(config).length`) to determine whether should run or not, and apparently is does not run. ;)

Therefore always assert things that should be there, and if things shouldn't be there assert that. This will make your tests more clear, shorter and bullet proof. Or we'll end up having tests for tests ;)

Reply:
On my TODO list

Reply:
Good catch! Fixed :)